contrib/mesos/pkg/scheduler/resource/resource_test.go

```diff
@@ -0,0 +1,111 @@
+		{2.0, 3.0, nan, 768.0, requestLimits{defCpu + 2.0, defCpu + 3.0, defMem + 768.0, defMem + 768.0}},
+		{2.0, 3.0, 256.0, 768.0, requestLimits{defCpu + 2.0, defCpu + 3.0, defMem + 256.0, defMem + 768.0}},
+	} {
+		pod := &api.Pod{
+			Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "a",
+				}, {
+					Name: "b",
+				}},
+			},
+		}
+
+		if !math.IsNaN(t.cpuR) || !math.IsNaN(t.memR) {
+			pod.Spec.Containers[0].Resources.Requests = api.ResourceList{}
+			if !math.IsNaN(t.cpuR) {
+				pod.Spec.Containers[0].Resources.Requests[api.ResourceCPU] = *CPUShares(t.cpuR).Quantity()
+			}
+			if !math.IsNaN(t.memR) {
+				pod.Spec.Containers[0].Resources.Requests[api.ResourceMemory] = *MegaBytes(t.memR).Quantity()
+			}
+		}
+		if !math.IsNaN(t.cpuL) || !math.IsNaN(t.memL) {
+			pod.Spec.Containers[0].Resources.Limits = api.ResourceList{}
+			if !math.IsNaN(t.cpuL) {
+				pod.Spec.Containers[0].Resources.Limits[api.ResourceCPU] = *CPUShares(t.cpuL).Quantity()
+			}
+			if !math.IsNaN(t.memL) {
+				pod.Spec.Containers[0].Resources.Limits[api.ResourceMemory] = *MegaBytes(t.memL).Quantity()
+			}
+		}
+
+		tst.Logf("Testing resource computation for %v => request=%v limit=%v", t, pod.Spec.Containers[0].Resources.Requests, pod.Spec.Containers[0].Resources.Limits)
+		tst.Logf("hasRequests: cpu => %v, mem => %v", resourcequota.PodHasRequests(pod, api.ResourceCPU), resourcequota.PodHasRequests(pod, api.ResourceMemory))
+
+		beforeCpuR, beforeCpuL, _, err := CPUForPod(pod, DefaultDefaultContainerCPULimit)
+		assert.NoError(err, "CPUForPod should not return an error")
+
+		beforeMemR, beforeMemL, _, err := MemForPod(pod, DefaultDefaultContainerMemLimit)
+		assert.NoError(err, "MemForPod should not return an error")
+
+		cpuR, cpuL, _, err := LimitPodCPU(pod, DefaultDefaultContainerCPULimit)
+		assert.NoError(err, "LimitPodCPU should not return an error")
+
+		memR, memL, _, err := LimitPodMem(pod, DefaultDefaultContainerMemLimit)
+		assert.NoError(err, "LimitPodMem should not return an error")
+
+		tst.Logf("New resources container 0: request=%v limit=%v", pod.Spec.Containers[0].Resources.Requests, pod.Spec.Containers[0].Resources.Limits)
+		tst.Logf("New resources container 1: request=%v limit=%v", pod.Spec.Containers[1].Resources.Requests, pod.Spec.Containers[1].Resources.Limits)
+
+		assert.Equal(t.want.cpuR, float64(beforeCpuR), "cpu request before modifiation is wrong")
+		assert.Equal(t.want.cpuL, float64(beforeCpuL), "cpu limit before modifiation is wrong")
+
+		assert.Equal(t.want.memR, float64(beforeMemR), "mem request before modifiation is wrong")
+		assert.Equal(t.want.memL, float64(beforeMemL), "mem limit before modifiation is wrong")
+
+		assert.Equal(t.want.cpuR, float64(cpuR), "cpu request is wrong")
+		assert.Equal(t.want.cpuL, float64(cpuL), "cpu limit is wrong")
+
+		assert.Equal(t.want.memR, float64(memR), "mem request is wrong")
+		assert.Equal(t.want.memL, float64(memL), "mem limit is wrong")
+	}
```

Comment:
I guess I was thinking that it might be nice to test the invariant that is
supposed to be guaranteed by `podResources`: limit >= requested. Perhaps
that's implicit given the current set of test data.

On Wed, Nov 11, 2015 at 7:41 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/resource/resource_test.go
> <https://github.com/kubernetes/kubernetes/pull/17054#discussion_r44528080>
> :
>
> > +
> > +		tst.Logf("New resources container 0: request=%v limit=%v", pod.Spec.Containers[0].Resources.Requests, pod.Spec.Containers[0].Resources.Limits)
> > +		tst.Logf("New resources container 1: request=%v limit=%v", pod.Spec.Containers[1].Resources.Requests, pod.Spec.Containers[1].Resources.Limits)
> > +
> > +		assert.Equal(t.want.cpuR, float64(beforeCpuR), "cpu request before modifiation is wrong")
> > +		assert.Equal(t.want.cpuL, float64(beforeCpuL), "cpu limit before modifiation is wrong")
> > +
> > +		assert.Equal(t.want.memR, float64(beforeMemR), "mem request before modifiation is wrong")
> > +		assert.Equal(t.want.memL, float64(beforeMemL), "mem limit before modifiation is wrong")
> > +
> > +		assert.Equal(t.want.cpuR, float64(cpuR), "cpu request is wrong")
> > +		assert.Equal(t.want.cpuL, float64(cpuL), "cpu limit is wrong")
> > +
> > +		assert.Equal(t.want.memR, float64(memR), "mem request is wrong")
> > +		assert.Equal(t.want.memL, float64(memL), "mem limit is wrong")
> > +	}
>
> We could add such a test to the podResources function. In the test it does
> not make much sense, because it would only test the values given by the
> test data.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17054/files#r44528080>.
>